Class list support for Completion#type

ColinTimBarndt opened the pull request:
This pull request allows supplying the `Completion#type` with an array of class names that are added to the icon `<div>` element without a prefix. This allows more CSS styling using selectors.

Example: You want an icon with decorations (a Java class can be final or static etc.). To avoid having to create a new class for every possible combination of modifiers and types, it would now be possible to add a class for each modifier, like "icon-final" or "icon-static" that can be combined: `{ label: "MyClass", type: ["icon", "type-icon", "icon-class", "icon-final"] }`.

I was unable to actually test this PR because the cm-buildhelper throws an internal error for me.

marijnh commented:
This is a good idea. A few nitpicks though:

 - I'd prefer to use a space-separated string rather than an array here, to keep the type simple.

 - Since doc comments are interpreted as markdown, you'll need an empty line (`///`) to start a new paragraph.

 - I generally appreciate minimalism in code but please do wrap nested if statements in braces (though with a space-separated string you probably don't need an `if` there anymore)

ColinTimBarndt commented:
>     * I'd prefer to use a space-separated string rather than an array here, to keep the type simple.

All classes should be prefixed in that case. It would be counterintuitive if one class (string) was prefixed and a list of classes (as a string) was not. What would be a good way to implement a feature that allows to disable the prefix? Should a new `classPrefix?: string` option be added?

marijnh commented:
Yes, all should definitely be prefixed—something like `.type.split(" ").map(n => "cm-completionIcon-" + n).join(" ")`

ColinTimBarndt commented:
I have fixed the issues you mentioned and added a new `typeClassPrefix` option which allows changing the prefix if really needed (for example using an external icon CSS library or using icon styles that are also used outside of codemirror).

<details><summary>Example</summary>

![screenshot](https://user-images.githubusercontent.com/26167910/120514723-e3a6bf00-c3cd-11eb-8933-5b7aae6d5895.png)

```typescript
export const javaAutocomplete = javaLanguage.data.of({
	autocomplete({ pos, state }: CompletionContext): CompletionResult | Promise<CompletionResult> | null {
		const top = syntaxTree(state);
		analyzeTree(state.doc, top);
		
		return {
			from: pos,
			options: [
				{ label: "TestClass", type: "icon types-icon icon-class", typeClassPrefix: "" },
				{ label: "TestClassFinal", type: "icon types-icon icon-class icon-final", typeClassPrefix: "" },
				{ label: "TestClassStatic", type: "icon types-icon icon-class icon-static", typeClassPrefix: "" },
				{ label: "pubMethod", type: "icon types-icon icon-method icon-public", typeClassPrefix: "" }
			]
		}
	}
});
```
</details>

marijnh commented:
Do you need the configurable prefix? If not, I'd rather avoid it for the sake of minimalism.

ColinTimBarndt commented:
I have a project where I want to use icons both inside of CodeMirror and in other places. The icons are defined in a generated sprite map and a stylesheet that has all the selectors for all the combinations. There are already a lot of selectors in that file and if I had to add the prefix support, I would have to double all selectors:

`.icon.types-icon.icon-class.icon-private.icon-final.icon-deprecated` would become `.cm-completionIcon-types-icon.cm-completionIcon-class.cm-completionIcon-private.cm-completionIcon-final.cm-completionIcon-deprecated, .icon.types-icon.icon-class.icon-private.icon-final.icon-deprecated`

If someone wants to use an external CSS icons library, they would need to rewrite it to add all the prefixes to the selectors. An example would be [Fontawesome](https://fontawesome.com/v6.0/icons).

ColinTimBarndt commented:
Another option I would have (because I wrote the CSS and Spritemap generator) is to generate abstract class names like `cm-completionIcon-types-a`, b, c etc. for which I would generate a json file as a map. I could do this for my code because I created the icons myself. This would not easily be possible for external icons.

marijnh commented:
Good points, but I kind of don't want to be _too_ flexible in cases like this, because at some point it becomes hard to write themes if there are few guarantees about what the DOM will look like. So I'd prefer to keep these class prefixes static (just like most of the other classes assigned by the library).

ColinTimBarndt commented:
Okay, I'm going to remove the feature. I might fork this repo to make a customized version with a different DOM structure to suit my needs better. I agree that simplicity is better in 99% of the cases and that most users of this module probably won't build an entire Web IDE using CodeMirror.

Is this PR okay otherwise?

marijnh commented:
Yes, thanks, looks great now.
